File: backend/app/routers/teams.py
from fastapi import APIRouter, Depends, HTTPException

from app.config import Settings
from app.dependencies import get_current_user, get_settings
from app.models.team import Team, TeamMembership
from app.models.user import User
from app.schemas.teams import (
    ChangeRoleRequest,
    CreateJoinLinkRequest,
    CreateTeamRequest,
    InviteRequest,
    RemoveMemberRequest,
    TransferOwnershipRequest,
    UpdateTeamNameRequest,
)
from app.services import team_service

router = APIRouter()


@router.get("/")
async def list_teams(user: User = Depends(get_current_user)):
    """List all teams the user belongs to."""
    return await team_service.get_user_teams(user.user_id)


@router.get("/{team_uuid}/members")
async def list_members(team_uuid: str, user: User = Depends(get_current_user)):
    """List members of a team."""
    team = await Team.find_one(Team.uuid == team_uuid)
    if not team:
        raise HTTPException(status_code=404, detail="Team not found")
    membership = await TeamMembership.find_one(
        TeamMembership.team == team.id,
        TeamMembership.user_id == user.user_id,
    )
    if not membership and not user.is_admin:
        raise HTTPException(status_code=403, detail="Not a member of this team")
    return await team_service.get_team_members(team.id)


@router.get("/{team_uuid}/invites")
async def list_invites(team_uuid: str, user: User = Depends(get_current_user)):
    """List pending invites for a team."""
    team = await Team.find_one(Team.uuid == team_uuid)
    if not team:
        raise HTTPException(status_code=404, detail="Team not found")
    membership = await TeamMembership.find_one(
        TeamMembership.team == team.id,
        TeamMembership.user_id == user.user_id,
    )
    if not membership and not user.is_admin:
        raise HTTPException(status_code=403, detail="Not a member of this team")
    if membership and membership.role not in ("owner", "admin") and not user.is_admin:
        raise HTTPException(status_code=403, detail="Admin or owner role required")
    return await team_service.get_team_invites(team.id)


@router.post("/create")
async def create_team(
    body: CreateTeamRequest,
    user: User = Depends(get_current_user),
):
    team = await team_service.create_team(body.name, user.user_id)
    return {"id": str(team.id), "uuid": team.uuid, "name": team.name}


@router.patch("/update_name")
async def update_name(
    body: UpdateTeamNameRequest,
    user: User = Depends(get_current_user),
):
    try:
        team = await team_service.update_team_name(
            body.team_id, body.name, user.user_id
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"status": "success", "name": team.name}


@router.post("/invite")
async def invite(
    body: InviteRequest,
    user: User = Depends(get_current_user),
    settings: Settings = Depends(get_settings),
):
    try:
        inv = await team_service.invite_member(
            body.team_id, body.email, body.role, user.user_id,
            settings=settings,
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"token": inv.token, "email": inv.email}


@router.get("/invite/info/{token}")
async def invite_info(token: str):
    """Public — return team + inviter metadata for a pending invite token."""
    info = await team_service.get_invite_info(token)
    if not info:
        raise HTTPException(status_code=404, detail="Invalid or already-used invite")
    return info


@router.post("/invite/accept/{token}")
async def accept_invite(
    token: str,
    user: User = Depends(get_current_user),
):
    try:
        team = await team_service.accept_invite(token, user)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"uuid": team.uuid, "name": team.name}


@router.post("/{team_uuid}/join-link")
async def create_join_link(
    team_uuid: str,
    body: CreateJoinLinkRequest,
    user: User = Depends(get_current_user),
):
    """Create a new public join link for the team. Admin/owner only."""
    try:
        link = await team_service.create_join_link(
            team_uuid,
            user.user_id,
            role=body.role,
            expires_in_hours=body.expires_in_hours,
            max_uses=body.max_uses,
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return team_service._serialize_join_link(link)


@router.get("/{team_uuid}/join-links")
async def list_join_links(
    team_uuid: str,
    user: User = Depends(get_current_user),
):
    """List active (non-revoked) join links for a team. Admin/owner only."""
    try:
        return await team_service.get_team_join_links(team_uuid, user.user_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.delete("/join-link/{token}")
async def revoke_join_link(
    token: str,
    user: User = Depends(get_current_user),
):
    """Revoke a join link. Admin/owner of the team only."""
    try:
        await team_service.revoke_join_link(token, user.user_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"ok": True}


@router.get("/join-link/info/{token}")
async def join_link_info(token: str):
    """Public — return team metadata for a join-link token."""
    info = await team_service.get_join_link_info(token)
    if not info:
        raise HTTPException(status_code=404, detail="Invalid join link")
    return info


@router.post("/join-link/accept/{token}")
async def accept_join_link(
    token: str,
    user: User = Depends(get_current_user),
):
    """Accept a join link — adds the current user to the team."""
    try:
        team = await team_service.accept_join_link(token, user)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"uuid": team.uuid, "name": team.name}


@router.post("/switch/{team_uuid}")
async def switch_team(
    team_uuid: str,
    user: User = Depends(get_current_user),
):
    try:
        team = await team_service.switch_team(team_uuid, user)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"uuid": team.uuid, "name": team.name}


@router.post("/member/role")
async def change_role(
    body: ChangeRoleRequest,
    user: User = Depends(get_current_user),
):
    try:
        await team_service.change_role(
            body.team_id, body.user_id, body.role, user.user_id
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"ok": True}


@router.post("/member/remove")
async def remove_member(
    body: RemoveMemberRequest,
    user: User = Depends(get_current_user),
):
    try:
        await team_service.remove_member(
            body.team_id, body.user_id, user.user_id
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"ok": True}


@router.post("/transfer-ownership")
async def transfer_ownership(
    body: TransferOwnershipRequest,
    user: User = Depends(get_current_user),
):
    try:
        team = await team_service.transfer_ownership(
            body.team_uuid, user.user_id, body.new_owner_user_id
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"uuid": team.uuid, "name": team.name, "owner_user_id": team.owner_user_id}


@router.delete("/{team_uuid}")
async def delete_team(
    team_uuid: str,
    user: User = Depends(get_current_user),
):
    try:
        await team_service.delete_team(team_uuid, user.user_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"ok": True}

File: frontend/src/api/teams.ts
import { apiFetch } from './client'
import type { Team, TeamMember, TeamInvite, TeamJoinLink } from '../types/user'

export function listTeams() {
  return apiFetch<Team[]>('/api/teams/')
}

export function createTeam(name: string) {
  return apiFetch<{ id: string; uuid: string; name: string }>('/api/teams/create', {
    method: 'POST',
    body: JSON.stringify({ name }),
  })
}

export function updateTeamName(team_id: string, name: string) {
  return apiFetch<{ status: string; name: string }>('/api/teams/update_name', {
    method: 'PATCH',
    body: JSON.stringify({ team_id, name }),
  })
}

export function switchTeam(teamUuid: string) {
  return apiFetch<{ uuid: string; name: string }>(`/api/teams/switch/${teamUuid}`, {
    method: 'POST',
  })
}

export function getTeamMembers(teamUuid: string) {
  return apiFetch<TeamMember[]>(`/api/teams/${teamUuid}/members`)
}

export function getTeamInvites(teamUuid: string) {
  return apiFetch<TeamInvite[]>(`/api/teams/${teamUuid}/invites`)
}

export function inviteMember(team_id: string, email: string, role: string = 'member') {
  return apiFetch<{ token: string; email: string }>('/api/teams/invite', {
    method: 'POST',
    body: JSON.stringify({ team_id, email, role }),
  })
}

export function acceptInvite(token: string) {
  return apiFetch<{ uuid: string; name: string }>(`/api/teams/invite/accept/${token}`, {
    method: 'POST',
  })
}

export interface InviteInfo {
  email: string
  role: string
  team_name: string
  team_uuid: string | null
  inviter_name: string | null
  expired: boolean
}

export async function getInviteInfo(token: string): Promise<InviteInfo> {
  const res = await fetch(`/api/teams/invite/info/${encodeURIComponent(token)}`)
  if (!res.ok) {
    throw new Error('Invalid or expired invite link.')
  }
  return res.json()
}

export function changeMemberRole(team_id: string, user_id: string, role: string) {
  return apiFetch<{ ok: boolean }>('/api/teams/member/role', {
    method: 'POST',
    body: JSON.stringify({ team_id, user_id, role }),
  })
}

export function removeMember(team_id: string, user_id: string) {
  return apiFetch<{ ok: boolean }>('/api/teams/member/remove', {
    method: 'POST',
    body: JSON.stringify({ team_id, user_id }),
  })
}

export function transferOwnership(teamUuid: string, newOwnerUserId: string) {
  return apiFetch<{ ok: boolean }>('/api/teams/transfer-ownership', {
    method: 'POST',
    body: JSON.stringify({ team_uuid: teamUuid, new_owner_user_id: newOwnerUserId }),
  })
}

export function deleteTeam(teamUuid: string) {
  return apiFetch<{ ok: boolean }>(`/api/teams/${teamUuid}`, {
    method: 'DELETE',
  })
}

// ---------------------------------------------------------------------------
// Public join links
// ---------------------------------------------------------------------------

export interface CreateJoinLinkParams {
  role?: string
  expires_in_hours?: number
  max_uses?: number | null
}

export function createJoinLink(teamUuid: string, params: CreateJoinLinkParams = {}) {
  return apiFetch<TeamJoinLink>(`/api/teams/${teamUuid}/join-link`, {
    method: 'POST',
    body: JSON.stringify({
      role: params.role ?? 'member',
      expires_in_hours: params.expires_in_hours ?? 48,
      max_uses: params.max_uses ?? null,
    }),
  })
}

export function getJoinLinks(teamUuid: string) {
  return apiFetch<TeamJoinLink[]>(`/api/teams/${teamUuid}/join-links`)
}

export function revokeJoinLink(token: string) {
  return apiFetch<{ ok: boolean }>(
    `/api/teams/join-link/${encodeURIComponent(token)}`,
    { method: 'DELETE' },
  )
}

export interface JoinLinkInfo {
  role: string
  team_name: string
  team_uuid: string | null
  inviter_name: string | null
  expires_at: string | null
  status: null | 'revoked' | 'expired' | 'exhausted'
}

export async function getJoinLinkInfo(token: string): Promise<JoinLinkInfo> {
  const res = await fetch(`/api/teams/join-link/info/${encodeURIComponent(token)}`)
  if (!res.ok) {
    throw new Error('Invalid join link.')
  }
  return res.json()
}

export function acceptJoinLink(token: string) {
  return apiFetch<{ uuid: string; name: string }>(
    `/api/teams/join-link/accept/${encodeURIComponent(token)}`,
    { method: 'POST' },
  )
}
